File: miniapp/src/utils/clubHours.ts
import type { Locale } from "../i18n";
import type { ClubOpenStatus } from "../types/api";

const DAY_ABBR: Record<string, number> = {
  пн: 0,
  вт: 1,
  ср: 2,
  чт: 3,
  пт: 4,
  сб: 5,
  вс: 6,
};

const LINE_PATTERN =
  /([ПпВвСсЧч][нтрбс]?)(?:[–\-—]([ПпВвСсЧч][нтрбс]?))?\s*:\s*(\d{1,2}):(\d{2})\s*[–\-—]\s*(\d{1,2}):(\d{2})/;

interface DaySchedule {
  weekday: number;
  openMinutes: number;
  closeMinutes: number;
}

function weekdayFromAbbr(value: string): number | null {
  const key = value.trim().toLowerCase().slice(0, 2);
  return DAY_ABBR[key] ?? null;
}

function expandWeekdays(start: number, end: number | null): number[] {
  if (end === null) {
    return [start];
  }
  if (start <= end) {
    return Array.from({ length: end - start + 1 }, (_, index) => start + index);
  }
  return [
    ...Array.from({ length: 7 - start }, (_, index) => start + index),
    ...Array.from({ length: end + 1 }, (_, index) => index),
  ];
}

function parseTimeToMinutes(hour: string, minute: string): number {
  return Number(hour) * 60 + Number(minute);
}

function formatMinutes(totalMinutes: number): string {
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}`;
}

function formatWeekdayShort(weekday: number, locale: Locale): string {
  const monday = new Date(Date.UTC(2024, 0, 1));
  const date = new Date(monday);
  date.setUTCDate(monday.getUTCDate() + weekday);
  const intlLocale = locale === "ua" ? "uk" : locale;
  return new Intl.DateTimeFormat(intlLocale, { weekday: "short", timeZone: "UTC" }).format(date);
}

export function parseHoursText(hours: string): DaySchedule[] {
  const schedules: DaySchedule[] = [];

  for (const line of hours.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) {
      continue;
    }

    const match = LINE_PATTERN.exec(trimmed);
    if (!match) {
      continue;
    }

    const startDay = weekdayFromAbbr(match[1]);
    if (startDay === null) {
      continue;
    }

    const endDay = match[2] ? weekdayFromAbbr(match[2]) : null;
    if (match[2] && endDay === null) {
      continue;
    }

    const openMinutes = parseTimeToMinutes(match[3], match[4]);
    const closeMinutes = parseTimeToMinutes(match[5], match[6]);

    for (const weekday of expandWeekdays(startDay, endDay)) {
      schedules.push({ weekday, openMinutes, closeMinutes });
    }
  }

  return schedules;
}

export function computeOpenStatus(
  hours: string,
  override?: string | null,
  timezone = "Europe/Kyiv",
  locale: Locale = "ru",
): ClubOpenStatus {
  if (override?.trim()) {
    return { is_open: null, label: override.trim() };
  }

  const schedules = parseHoursText(hours);
  if (!schedules.length) {
    return { is_open: null, messageKey: "club.status.hoursUnknown" };
  }

  const now = new Date();
  const formatter = new Intl.DateTimeFormat("en-GB", {
    timeZone: timezone,
    weekday: "short",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  });
  const parts = formatter.formatToParts(now);
  const weekdayPart = parts.find((part) => part.type === "weekday")?.value ?? "Mon";
  const hour = Number(parts.find((part) => part.type === "hour")?.value ?? "0");
  const minute = Number(parts.find((part) => part.type === "minute")?.value ?? "0");

  const weekdayMap: Record<string, number> = {
    Mon: 0,
    Tue: 1,
    Wed: 2,
    Thu: 3,
    Fri: 4,
    Sat: 5,
    Sun: 6,
  };
  const today = weekdayMap[weekdayPart] ?? 0;
  const currentMinutes = hour * 60 + minute;

  const todaySchedules = schedules
    .filter((item) => item.weekday === today)
    .sort((left, right) => left.openMinutes - right.openMinutes);

  for (const item of todaySchedules) {
    if (item.openMinutes <= currentMinutes && currentMinutes <= item.closeMinutes) {
      return {
        is_open: true,
        messageKey: "club.status.openUntil",
        messageVars: { time: formatMinutes(item.closeMinutes) },
      };
    }
  }

  for (const item of todaySchedules) {
    if (currentMinutes < item.openMinutes) {
      return {
        is_open: false,
        messageKey: "club.status.opensAt",
        messageVars: { time: formatMinutes(item.openMinutes) },
      };
    }
  }

  for (let offset = 1; offset < 8; offset += 1) {
    const weekday = (today + offset) % 7;
    const daySchedules = schedules
      .filter((item) => item.weekday === weekday)
      .sort((left, right) => left.openMinutes - right.openMinutes);

    if (!daySchedules.length) {
      continue;
    }

    const nextOpen = daySchedules[0];
    const time = formatMinutes(nextOpen.openMinutes);
    if (offset === 1) {
      return {
        is_open: false,
        messageKey: "club.status.opensTomorrow",
        messageVars: { time },
      };
    }

    return {
      is_open: false,
      messageKey: "club.status.opensOnDay",
      messageVars: { day: formatWeekdayShort(weekday, locale), time },
    };
  }

  return { is_open: false, messageKey: "club.status.closed" };
}

export function formatOpenStatusLabel(
  status: ClubOpenStatus,
  t: (path: string, vars?: Record<string, string | number>) => string,
): string {
  if (status.label) {
    return status.label;
  }
  if (status.messageKey) {
    return t(status.messageKey, status.messageVars);
  }
  return "";
}

File: shared/bgclub/services/club_hours.py
import re
from dataclasses import dataclass
from datetime import datetime, time, timezone
from zoneinfo import ZoneInfo

DAY_ABBR = {
    "пн": 0,
    "вт": 1,
    "ср": 2,
    "чт": 3,
    "пт": 4,
    "сб": 5,
    "вс": 6,
}

DAY_LABELS = ["пн", "вт", "ср", "чт", "пт", "сб", "вс"]

LINE_PATTERN = re.compile(
    r"([ПпВвСсЧч][нтрбс]?)"
    r"(?:[–\-—]([ПпВвСсЧч][нтрбс]?))?"
    r"\s*:\s*"
    r"(\d{1,2}):(\d{2})\s*[–\-—]\s*(\d{1,2}):(\d{2})"
)


@dataclass(frozen=True)
class DaySchedule:
    weekday: int
    open_time: time
    close_time: time


@dataclass(frozen=True)
class OpenStatus:
    is_open: bool | None
    label: str


def _weekday_from_abbr(value: str) -> int | None:
    key = value.strip().lower()[:2]
    return DAY_ABBR.get(key)


def _expand_weekdays(start: int, end: int | None) -> list[int]:
    if end is None:
        return [start]
    if start <= end:
        return list(range(start, end + 1))
    return list(range(start, 7)) + list(range(0, end + 1))


def parse_hours_text(hours: str) -> list[DaySchedule]:
    schedules: list[DaySchedule] = []

    for line in hours.splitlines():
        line = line.strip()
        if not line:
            continue

        match = LINE_PATTERN.search(line)
        if not match:
            continue

        start_day = _weekday_from_abbr(match.group(1))
        if start_day is None:
            continue

        end_day = _weekday_from_abbr(match.group(2)) if match.group(2) else None
        if match.group(2) and end_day is None:
            continue

        open_time = time(int(match.group(3)), int(match.group(4)))
        close_time = time(int(match.group(5)), int(match.group(6)))

        for weekday in _expand_weekdays(start_day, end_day):
            schedules.append(
                DaySchedule(weekday=weekday, open_time=open_time, close_time=close_time)
            )

    return schedules


def compute_open_status(
    schedules: list[DaySchedule],
    *,
    now: datetime | None = None,
    override: str | None = None,
    club_timezone: str = "Europe/Kyiv",
) -> OpenStatus:
    if override and override.strip():
        return OpenStatus(is_open=None, label=override.strip())

    if not schedules:
        return OpenStatus(is_open=None, label="Часы работы не указаны")

    moment = now or datetime.now(timezone.utc)
    local_now = moment.astimezone(ZoneInfo(club_timezone))
    today = local_now.weekday()
    current_time = local_now.time()

    today_schedules = sorted(
        [item for item in schedules if item.weekday == today],
        key=lambda item: item.open_time,
    )

    for item in today_schedules:
        if item.open_time <= current_time <= item.close_time:
            return OpenStatus(
                is_open=True,
                label=f"Открыто до {item.close_time.strftime('%H:%M')}",
            )

    for item in today_schedules:
        if current_time < item.open_time:
            return OpenStatus(
                is_open=False,
                label=f"Откроемся в {item.open_time.strftime('%H:%M')}",
            )

    for offset in range(1, 8):
        weekday = (today + offset) % 7
        day_schedules = sorted(
            [item for item in schedules if item.weekday == weekday],
            key=lambda item: item.open_time,
        )
        if not day_schedules:
            continue

        next_open = day_schedules[0]
        if offset == 1:
            return OpenStatus(
                is_open=False,
                label=f"Откроемся завтра в {next_open.open_time.strftime('%H:%M')}",
            )
        return OpenStatus(
            is_open=False,
            label=(
                f"Откроемся в {DAY_LABELS[weekday]} "
                f"в {next_open.open_time.strftime('%H:%M')}"
            ),
        )

    return OpenStatus(is_open=False, label="Закрыто")
